File: web/test/integration/test/LoginSpec.js
'use strict'

const asUser = require('../support/asUser')

class LoginShortcuts {
  constructor(browser) {
    this.b = browser
  }

  async assertLoggedInAs(email) {
    await this.b.assertExists({ tag: 'div', class: 'logged-in', contains: email, wait: 'pageLoad' })
  }

  async assertLogInFailed() {
    await this.b.assertExists({ tag: 'h1', contains: 'Log in', wait: 'pageLoad' })
    await this.b.assertExists({ class: 'error', contains: 'Wrong email address or password' })
  }

  async assertLoggedOut() {
    await this.b.assertExists({ class: 'session-form', wait: 'pageLoad' })
  }

  async tryLogIn(email, password) {
    await this.b.sendKeys(email, '.session-form [name=email]')
    await this.b.sendKeys(password, '.session-form [name=password]')
    await this.b.click('.session-form [type=submit]')
  }

  async logOut() {
    await this.b.click({ link: 'Log out' })
    await this.b.assertExists({ class: 'session-form', wait: 'pageLoad' })
  }
}

// In order to let users store their own information
// We need to let users identify themselves
describe('Login', function() {
  asUser.usingTemporaryUser(function() {
    before(async function() {
      this.login = new LoginShortcuts(this.browser)
      this.b = this.browser

      // We'll start this test suite logged out.
      await this.login.logOut()
    })

    it('should log in', async function() {
      await this.login.tryLogIn(this.userEmail, this.userEmail)
      await this.login.assertLoggedInAs(this.userEmail)
      await this.login.logOut()
    })

    it('should log out', async function() {
      // This is easy -- we logged out in the before hook ;)
      await this.login.assertLoggedOut()
    })

    it('should not log in with wrong password', async function() {
      await this.login.tryLogIn(this.userEmail, 'bad-password')
      await this.login.assertLogInFailed()
    })

    it('should not log in with the wrong username', async function() {
      await this.login.tryLogIn("x-#{this.userEmail}", this.userEmail)
      await this.login.assertLogInFailed()
    })
  })
})

File: integration-test/spec/login_spec.rb
#!/usr/bin/env ruby

require './spec/spec_helper'

# In order to let users store their own information
# We need to let users identify themselves
describe 'Login' do
  before do
    @user = admin_session.create_test_user
  end

  after do
    admin_session.destroy_test_user(@user)
  end

  def assert_logged_in_as(email)
    page.assert_selector('div.logged-in', text: email, wait: WAIT_LOAD)
  end

  def assert_log_in_failed
    page.assert_selector('h1', text: /Log in/i, wait: WAIT_LOAD)
    page.assert_selector('.error', text: 'Wrong email address or password')
  end

  def assert_logged_out
    page.assert_selector('.session-form', wait: WAIT_LOAD)
  end

  def try_log_in(email, password)
    page.visit('/')
    page.within('.session-form') do
      page.fill_in('email', with: email)
      page.fill_in('password', with: password)
      page.click_button('Log in')
    end
  end

  def log_out
    page.click_link('Log out')
  end

  it 'should log in' do
    try_log_in(@user[:email], @user[:password])
    assert_logged_in_as(@user[:email])
  end

  it 'should log out' do
    try_log_in(@user[:email], @user[:password])
    log_out
    assert_logged_out
  end

  it 'should not log in with wrong password' do
    try_log_in(@user[:email], @user[:password] + 'x')
    assert_log_in_failed
  end

  it 'should not log in with wrong username' do
    try_log_in('x' + @user[:email], @user[:password])
    assert_log_in_failed
  end
end
